### PR TITLE
Fixes #8216: Add a rudder-dev merge pr to merge a specific PR

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/8216

### CODE REVIEW

**Comment by peckpeck** [comment]
PR updated

File: scripts/rudder-dev/rudder-dev [hunk 1]
```diff
@@ -20,8 +20,9 @@ Usage:
     rudder-dev [-f|--force] takeover <ticket_id>
     rudder-dev [-f|--force] revert <ticket_id> [--retarget]
     rudder-dev [-f|--force] merge all [-s <strategy>] [-a|--automatic]
-    rudder-dev [-f|--force] merge <first_branch> [-s <strategy>] [-a|--automatic]
-    rudder-dev [-f|--force] merge <first_branch> <next_branch> [-s <strategy>] [-a|--automatic]
+    rudder-dev [-f|--force] merge <first_branch/ticket_id> [-s <strategy>] [-a|--automatic]
+    rudder-dev [-f|--force] merge <first_branch/ticket_id> <next_branch> [-s <strategy>] [-a|--automatic]
+    rudder-dev [-f|--force] merge [-s <strategy>] [-a|--automatic]
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: What does merge without param do ?  check current branch for branch / ticket id ?

File: scripts/rudder-dev/rudder-dev [hunk 2]
```diff
@@ -239,6 +242,138 @@ IN_PROGRESS_CODE = 9
...
+class Issue:
+  """Class to hold informations about a single issue"""
+  def __init__(self, name, must_be_open=True):
+    """name is a string like: 1234 or i1234"""
+    self.name = name
+    self.must_be_open = must_be_open
+    self.info = None
+    if name.startswith('#'):
+      name = name [1:]
+    match = re.match(r'i(\d+)', name)
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: Here they all need to be prefexied by i ?  Should the regexp not be "i?(\d)+" ?

Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: ok understood... match should be something else to me :)  something like "is_intern_issue"

File: scripts/rudder-dev/rudder-dev [hunk 3]
```diff
@@ -312,6 +447,7 @@ def get_version(ticket, error_fail=True):
  
 
 # Get informations about a ticket from redmine
+# Deprecated, use Issue()['field']
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: There is still some calls to get_ticket_info ?

Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: Yes, that's why it's just deprecated

File: scripts/rudder-dev/rudder-dev (hunk 2 above)
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: Yes a little rename could be useful

File: scripts/rudder-dev/rudder-dev (hunk 1 above)
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: exactly

**Comment by peckpeck** [comment]
PR updated

File: scripts/rudder-dev/rudder-dev (hunk 1 above)
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: Can you add this behavior in the desription of the command please ? :)

File: scripts/rudder-dev/rudder-dev [hunk 4]
```diff
@@ -1635,11 +1801,18 @@ if __name__ == "__main__":
     before_merge_branch = current_branch
     if arguments['all']:
       merge_all(arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
+    elif arguments['<first_branch/ticket_id>'] is None:
+      merge_issue(strategy=arguments['<strategy>'], automatic=arguments['-a'] or arguments['--automatic'])
+    elif arguments['<next_branch>'] is None:
+      if re.match(r'^\d{3,}$', arguments['<first_branch/ticket_id>']): # 3 digits -> ticket_id, otherwise it's a branch
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: This won't work on intern branch (ie i3540)

File: scripts/rudder-dev/rudder-dev [hunk 5]
```diff
@@ -1635,11 +1801,18 @@ if __name__ == "__main__":
...
+      else:
+        merge_to_next(arguments['<first_branch/ticket_id>'], arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
     else:
-      if arguments['<next_branch>'] is None:
-        merge_to_next(arguments['<first_branch>'], arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
+      if re.match(r'^\d{3,}$', arguments['<first_branch/ticket_id>']): # 3 digits -> ticket_id, otherwise it's a branch
+        merge_issue(issue_name=arguments['<first_branch/ticket_id>'], version=arguments['<next_branch>'], strategy=arguments['<strategy>'], automatic=arguments['-a'] or arguments['--automatic'])
       else:
-        merge(arguments['<first_branch>'], arguments['<next_branch>'], arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
+        merge_version(arguments['<first_branch|ticket_id>'], arguments['<next_branch>'], arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: <first_branch|ticket_id> should be <first_branch/ticket_id>

(note the / and not the |  )

File: scripts/rudder-dev/rudder-dev [hunk 6]
```diff
@@ -1635,11 +1801,18 @@ if __name__ == "__main__":
...
+    elif arguments['<next_branch>'] is None:
+      if re.match(r'^\d{3,}$', arguments['<first_branch/ticket_id>']): # 3 digits -> ticket_id, otherwise it's a branch
+        merge_issue(issue_name=arguments['<first_branch/ticket_id>'], strategy=arguments['<strategy>'], automatic=arguments['-a'] or arguments['--automatic'])
+      else:
+        merge_to_next(arguments['<first_branch/ticket_id>'], arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
     else:
-      if arguments['<next_branch>'] is None:
-        merge_to_next(arguments['<first_branch>'], arguments['<strategy>'], arguments['-a'] or arguments['--automatic'])
+      if re.match(r'^\d{3,}$', arguments['<first_branch/ticket_id>']): # 3 digits -> ticket_id, otherwise it's a branch
+        merge_issue(issue_name=arguments['<first_branch/ticket_id>'], version=arguments['<next_branch>'], strategy=arguments['<strategy>'], automatic=arguments['-a'] or arguments['--automatic'])
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: You don't need the parameter names herel (version is set contrary to previous calls! )

**Comment by peckpeck** [comment]
PR updated

File: scripts/rudder-dev/rudder-dev [hunk 7]
```diff
@@ -142,34 +143,40 @@ REVERT
 
 MERGE
```
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: Behavior without params not explained (look for issue in branch name)

**Comment by VinceMacBuche** [comment]
Apart the missing documentation, it's ok to me!

File: scripts/rudder-dev/rudder-dev (hunk 7 above)
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: In the action list :
- detect the ticket id from current branch if there is no parameter

Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: ok, misread that line!!! great!

**Pull request merged**